File: src/main/kotlin/moe/gensoukyo/gui/pages/UnEmbeddingPage.java
package moe.gensoukyo.gui.pages;

import me.wuxie.wakeshow.wakeshow.api.WuxieAPI;
import me.wuxie.wakeshow.wakeshow.ui.Container;
import me.wuxie.wakeshow.wakeshow.ui.WInventoryScreen;
import me.wuxie.wakeshow.wakeshow.ui.WxScreen;
import me.wuxie.wakeshow.wakeshow.ui.component.*;
import moe.gensoukyo.gui.config.MainConfig;
import moe.gensoukyo.gui.util.EmbeddingTools;
import moe.gensoukyo.gui.util.Pos;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class UnEmbeddingPage implements Page {
    private final String VERSION = (String) (MainConfig.INSTANCE.getConf().get("imageVersion"));
    private final String GUI_BACKGROUND = String.format("https://cdn.jsdelivr.net/gh/MineCraftGensoukyo/MCGImages@%s/img/Unembedding_BG.png",VERSION);
    private final String BTN_1 = String.format("https://cdn.jsdelivr.net/gh/MineCraftGensoukyo/MCGImages@%s/img/Unembedding_BTN_1.png",VERSION);
    private final String BTN_2 = String.format("https://cdn.jsdelivr.net/gh/MineCraftGensoukyo/MCGImages@%s/img/Unembedding_BTN_2.png",VERSION);
    private final String BTN_3 = String.format("https://cdn.jsdelivr.net/gh/MineCraftGensoukyo/MCGImages@%s/img/Unembedding_BTN_3.png",VERSION);

    private final Pos GUI_POS = new Pos(-1, -1, 190, 190, 0,0);
    private final WInventoryScreen gui = new WInventoryScreen("摘除镶嵌UI",
            GUI_BACKGROUND,
            GUI_POS.getDx(),GUI_POS.getDy(),GUI_POS.getW(),GUI_POS.getH(),15,GUI_POS.getH()-80);
    private final Container guiContainer = gui.getContainer();
    private final List<String> titleList = new ArrayList<>();{
        titleList.add("§9§l摘除镶嵌");
    }
    private final WScrollingContainer scroll = new WScrollingContainer(guiContainer,"choose_scroll",
            147,4,39,68,200);
    private final WTextList titleText = new WTextList(guiContainer,"title_list",titleList,78,5,60,20);
    private final WTextList tipsText = new WTextList(guiContainer,"tips_list",new ArrayList<>(),35,62,60,20);
    private final WSlot equipmentSlot = new WSlot(guiContainer,"equipment_slot", new ItemStack(Material.AIR),88,42);
    private final WButton decide_button = new WButton(guiContainer,"decide_button","",
            BTN_1,BTN_2,BTN_3,
            25,35);
    private final Container scrollContainer = scroll.getContainer();
    {
        tipsText.setScale(0.9);
        decide_button.setW(45);
        decide_button.setH(13);
        decide_button.setCanPress(false);
        decide_button.setTooltips(Collections.singletonList("摘除镶嵌"));

        decide_button.setFunction((t,pl) -> {
            ItemStack item = equipmentSlot.getItemStack();
            List<String> itemLore = item.getItemMeta().getLore();
            List<String> unembeddingList = new ArrayList<>();

            scrollContainer.getComponentMap().forEach((key,value) -> {
                if(!(value instanceof WCheckBox)) return;
                WCheckBox choose = (WCheckBox) value;
                if(choose.isSelect()) unembeddingList.add(key);
            });

            if(unembeddingList.isEmpty()){
                tipsText.setContent(Collections.singletonList("§c未选择"));
            }
            else {
                List<List> returnList = EmbeddingTools.unEmbedding(itemLore,unembeddingList);
                List<String> newLore = returnList.get(0);
                List<ItemStack> primordialStoneList = returnList.get(1);

                if(EmbeddingTools.inventoryPlentyFor(pl, primordialStoneList)){
                    ItemStack newEquipment = item.clone();
                    ItemMeta newMeta = item.getItemMeta().clone();
                    newMeta.setLore(newLore);
                    newEquipment.setItemMeta(newMeta);
                    equipmentSlot.setItemStack(newEquipment);

                    primordialStoneList.forEach(pl.getInventory()::addItem);

                    tipsText.setContent(Collections.singletonList("§a摘除成功"));

                    decide_button.setCanPress(false);
                    scrollContainer.getComponentMap().keySet().forEach(scrollContainer::remove);
                }
                else {
                    tipsText.setContent(Collections.singletonList("§c背包不足"));
                }
            }

            WuxieAPI.updateGui(pl);
        });

        equipmentSlot.setCanDrag(true);
        scroll.setBarWidth(5);
        scroll.setShowScrollBar(false);

        guiContainer.add(decide_button);
        guiContainer.add(titleText);
        guiContainer.add(tipsText);
        guiContainer.add(scroll);
        guiContainer.add(equipmentSlot);

    }

    @Override
    public WxScreen getPage() {
        return gui;
    }

    @Override
    public void showPage(Player player) {
        DefaultImpls.showPage(this,player);
    }

    @Override
    public void showCachePage(Player player) {
        DefaultImpls.showCachePage(this,player);
    }
}

File: src/main/kotlin/moe/gensoukyo/gui/pages/UnEmbeddingPage.kt
package moe.gensoukyo.gui.pages

import me.wuxie.wakeshow.wakeshow.api.WuxieAPI
import me.wuxie.wakeshow.wakeshow.ui.ClickFunction
import me.wuxie.wakeshow.wakeshow.ui.Component
import me.wuxie.wakeshow.wakeshow.ui.WInventoryScreen
import me.wuxie.wakeshow.wakeshow.ui.WxScreen
import me.wuxie.wakeshow.wakeshow.ui.component.*
import moe.gensoukyo.gui.config.MainConfig.conf
import moe.gensoukyo.gui.util.EmbeddingTools
import moe.gensoukyo.gui.util.Pos
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import java.util.function.Consumer

class UnEmbeddingPage : Page {
    private val GUI_BACKGROUND = "location:mcgproject:textures/gui/unembedding_bg.png"
    private val BTN_1 = "location:mcgproject:textures/gui/unembedding_btn_1.png"
    private val BTN_2 = "location:mcgproject:textures/gui/unembedding_btn_2.png"
    private val BTN_3 = "location:mcgproject:textures/gui/unembedding_btn_3.png"
    private val GUI_POS = Pos(-1, -1, 190, 190, 0, 0)
    private val gui = WInventoryScreen(
        "摘除镶嵌UI",
        GUI_BACKGROUND,
        GUI_POS.dx, GUI_POS.dy, GUI_POS.w, GUI_POS.h, 15, GUI_POS.h - 80
    )
    private val guiContainer = gui.container
    private val scroll = WScrollingContainer(
        guiContainer, "choose_scroll",
        147, 4, 39, 68, 200
    )
    private val titleText = WTextList(guiContainer, "title_list", listOf("§1§l摘 除 镶 嵌"), 65, 4, 60, 20)
    private val tipsText = WTextList(guiContainer, "tips_list", listOf(), 35, 62, 60, 20)
    private val equipmentSlot = WSlot(guiContainer, "equipment_slot", ItemStack(Material.AIR), 88, 42)
    private val decide_button = WButton(
        guiContainer, "decide_button", "",
        BTN_1, BTN_2, BTN_3,
        25, 35
    )

    override fun getPage(): WxScreen {
        tipsText.scale = 0.7
        equipmentSlot.emptyTooltips = listOf("§f请放入装备")
        decide_button.w = 45
        decide_button.h = 13
        decide_button.isCanPress = false
        decide_button.tooltips = listOf("§f摘除镶嵌")
        decide_button.function = ClickFunction { t: Int, pl: Player ->
            val scrollContainer = scroll.container
            val item = equipmentSlot.itemStack
            val itemLore = item.itemMeta!!.lore!!
            val unembeddingList: MutableList<String> = ArrayList()
            scrollContainer.componentMap.forEach { (key: String, value: Component?) ->
                if (value !is WCheckBox) return@forEach
                if (value.isSelect) unembeddingList.add(key)
            }
            if (unembeddingList.isEmpty()) {
                tipsText.content = listOf("§4§l未选择")
            } else {
                val returnList = EmbeddingTools.unEmbedding(itemLore, unembeddingList)
                val newLore = returnList.first
                val primordialStoneList = returnList.second
                if (EmbeddingTools.inventoryPlentyFor(pl, primordialStoneList)) {
                    val newEquipment = item.clone()
                    val newMeta = item.itemMeta!!.clone()
                    newMeta.lore = newLore
                    newEquipment.setItemMeta(newMeta)
                    equipmentSlot.itemStack = newEquipment
                    primordialStoneList.forEach(Consumer { itemStacks: ItemStack? -> pl.inventory.addItem(itemStacks) })
                    decide_button.isCanPress = false
                    scrollContainer.componentMap.keys.forEach(Consumer { component: String? ->
                        scrollContainer.remove(
                            component
                        )
                    })

                    EmbeddingTools.unEmbeddingCheck(newEquipment, guiContainer, pl)
                    tipsText.content = listOf("§a§l摘除成功")
                } else {
                    tipsText.content = listOf("§4§l背包不足")
                }
            }
            WuxieAPI.updateGui(pl)
        }
        equipmentSlot.isCanDrag = true
        scroll.barWidth = 5
        scroll.isShowScrollBar = false
        titleText.scale = 1.2
        guiContainer.add(decide_button)
        guiContainer.add(titleText)
        guiContainer.add(tipsText)
        guiContainer.add(scroll)
        guiContainer.add(equipmentSlot)

        return gui
    }
}